pkg/auth/ldaputil/query.go

```diff
@@ -0,0 +1,211 @@
+	return ldap.NewSearchRequest(
+		q.BaseDN,
+		int(q.Scope),
+		int(q.DerefAliases),
+		0, // allowed return size - indicates no limit
+		q.TimeLimit,
+		false, // not types only
+		q.Filter,
+		additionalAttributes,
+		nil, // no controls
+	)
+}
+
+// LDAPQueryOnAttribute encodes an LDAP query that conjoins two filters to extract a specific LDAP entry
+// This query is not self-sufficient and needs the value of the QueryAttribute to construct the final filter
+type LDAPQueryOnAttribute struct {
+	// Query retrieves entries from an LDAP server
+	LDAPQuery
+
+	// QueryAttribute is the attribute for a specific filter that, when conjoined with the common filter,
+	// retrieves the specific LDAP entry from the LDAP server. (e.g. "cn", when formatted with "aGroupName"
+	// and conjoined with "objectClass=groupOfNames", becomes (&(objectClass=groupOfNames)(cn=aGroupName))")
+	QueryAttribute string
+}
+
+// IdentifiyingLDAPQueryOptions holds a query and the attribute that identifies the entries that the query returns
+type IdentifiyingLDAPQueryOptions struct {
+	// Query retrieves entries from an LDAP server
+	LDAPQueryOnAttribute
+
+	// NameAttributes defines the attributes for the LDAP entries returned by the Query that will be interpreted
+	// as their names
+	NameAttributes []string
+}
+
+// NewIdentifiyingLDAPQueryOptions converts a user-provided LDAPQuery into a version we can use by parsing
+// the input and combining it with a set of name attributes
+func NewIdentifiyingLDAPQueryOptions(config api.LDAPQuery,
+	nameAttributes []string) (IdentifiyingLDAPQueryOptions, error) {
+
+	scope, err := DetermineLDAPScope(config.Scope)
+	if err != nil {
+		return IdentifiyingLDAPQueryOptions{}, err
+	}
+
+	derefAliases, err := DetermineDerefAliasesBehavior(config.DerefAliases)
+	if err != nil {
+		return IdentifiyingLDAPQueryOptions{}, err
+	}
+
+	return IdentifiyingLDAPQueryOptions{
+		LDAPQueryOnAttribute: LDAPQueryOnAttribute{
+			LDAPQuery: LDAPQuery{
+				BaseDN:       config.BaseDN,
+				Scope:        scope,
+				DerefAliases: derefAliases,
+				TimeLimit:    config.TimeLimit,
+				Filter:       config.Filter,
+			},
+			QueryAttribute: config.QueryAttribute,
+		},
+		NameAttributes: nameAttributes,
+	}, nil
+}
+
+// NewSearchRequest creates a new search request from the identifying query by internalizing the value of
+// the attribute to be filtered as well as any additional attributest that need to be recovereds
+func (o *IdentifiyingLDAPQueryOptions) NewSearchRequest(attributeValue string,
+	additionalAttributes []string) (*ldap.SearchRequest, error) {
+
+	allAttributes := util.NewStringSet(o.NameAttributes...)
+	allAttributes.Insert(additionalAttributes...)
+
+	if o.QueryAttribute == "DN" || o.QueryAttribute == "dn" {
+		if !strings.Contains(attributeValue, o.BaseDN) {
+			return nil, fmt.Errorf("search for entry with %s=%s would search outside of the base dn specified (dn=%s)",
```

Comment:
Create a particular error type for this.  See https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/validation.go#L32 as an example